### PR TITLE
Relaxing in restrictions.

### PR DESCRIPTION
**Pull request by MarkDuckworth**
- relaxing in restrictions
- port of https://github.com/firebase/firebase-android-sdk/pull/4231

### CODE REVIEW

File: packages/firestore/test/integration/api/query.test.ts [hunk 1]
```diff
@@ -1546,6 +1546,241 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+        query(
+          coll,
+          or(where('a', 'in', [2, 3]), where('b', 'array-contains', 3))
+        ),
+        'doc3',
+        'doc4',
+        'doc6'
+      );
+    });
+
```
Comment (ehsannas) [inline packages/firestore/test/integration/api/query.test.ts:1723]: same thing here

**Review by ehsannas — CHANGES_REQUESTED** [review]

File: packages/firestore/test/integration/api/validation.test.ts [hunk 2]
```diff
@@ -1055,19 +1025,6 @@ apiDescribe('Validation:', (persistence: boolean) => {
...
-        query(
-          collection(db, 'test'),
-          where('foo', 'in', [2, 3]),
-          where('foo', 'array-contains', 1),
-          where('foo', 'array-contains-any', [2])
-        )
-      ).to.throw(
-        "Invalid query. You cannot use 'array-contains-any' filters with 'in' filters."
-      );
-
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/validation.test.ts:1070]: fixed

File: packages/firestore/test/integration/api/query.test.ts [hunk 3]
```diff
@@ -1546,6 +1546,241 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+      // Two IN operations on different fields with conjunction.
+      await checkOnlineAndOfflineResultsMatch(
+        query(
+          coll,
+          and(where('a', 'in', [2, 3]), where('b', 'in', [0, 2])),
+          orderBy('a')
+        ),
+        'doc3'
+      );
+
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1584]: added

File: packages/firestore/test/integration/api/query.test.ts [hunk 4]
```diff
@@ -1546,6 +1546,241 @@ apiDescribe('Queries', (persistence: boolean) => {
...
+          coll,
+          or(where('a', 'in', [2, 3]), where('b', 'array-contains-any', [0, 7]))
+        ),
+        'doc1',
+        'doc3',
+        'doc4',
+        'doc6'
+      );
+    });
+
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/integration/api/query.test.ts:1655]: yikes. Fixed

File: packages/firestore/test/integration/api/validation.test.ts [hunk 5]
```diff
@@ -961,106 +961,123 @@ apiDescribe('Validation:', (persistence: boolean) => {
...
-      expect(() =>
-        query(
-          collection(db, 'test'),
-          where('foo', 'not-in', [1, 2]),
-          where('foo', 'not-in', [2, 3])
-        )
-      ).to.throw(
-        "Invalid query. You cannot use more than one 'not-in' filter."
-      );
+    validationIt.only(
```
Comment (ehsannas) [inline packages/firestore/test/integration/api/validation.test.ts:964]: remote `.only`

**Review by ehsannas — APPROVED** [review]

**Comment by MarkDuckworth** [comment]
Cherry picked to https://github.com/firebase/firebase-js-sdk/pull/7024.

**Pull request closed without merging**